python-api/gstswitch/process_monitor.py

```diff
@@ -0,0 +1,157 @@
+"""
+The ProcessMonitor runs a command in a background thread and monitors
+its output- It can block until the command prints a certain string and
+log the full output into a file
+"""
+
+from __future__ import absolute_import, print_function, unicode_literals
+
+import subprocess
+import logging
+import sys
+import os
+import time
+import select
+
+from .exception import ServerProcessError
+from .exception import MatchTimeoutError, MatchEofError, SelectError
+
+
+class ProcessMonitor(subprocess.Popen):
+    """Runs a command in a background-thread and monitors its output
+
+    Can block until the command prints a certain string and log the full
+    output into a file
+    """
+
+    def __init__(self, cmd, cmd_output_target=sys.stderr):
+        self.log = logging.getLogger('server-output-monitor')
+
+        # Logfile to write to
+        self._cmd_output_target = cmd_output_target
+
+        # Internal Buffer to search for matched when wait_for_output is called
+        self._buffer = ""
+
+        self.log.debug("starting subprocess")
+
+        try:
+            super(ProcessMonitor, self).__init__(
+                cmd,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+                bufsize=0,
+                shell=False)
+        except Exception as err:
+            raise ServerProcessError(err)
+
+        self.log.debug("subprocess successfully started")
+
+    def terminate(self):
```

Comment:
I guess you're right. I changed the code to something like what you suggested, although python2's subprocess-module does not offer timeouts and using subprocess32 here requires writing another set of unit-tests to test against subprocess32, so I just moved the terminatr before the read-cycle.